lib/journaled/abstract_adapter_ext.rb

```diff
@@ -0,0 +1,68 @@
+module Journaled
+  module AbstractAdapterExt
+    def self.included(klass)
+      klass.set_callback(:checkout, :before) do
+        @_journaled_transaction_handler = nil
+      end
+    end
+
+    def _journaled_transaction_handler
+      if @_journaled_transaction_handler&.active?
+        @_journaled_transaction_handler
+      else
+        @_journaled_transaction_handler = TransactionHandler.new(connection: self)
+      end
+    end
+
+    class TransactionHandler
+      def initialize(connection:)
+        raise TransactionSafetyError, <<~MSG unless connection.transaction_open?
+          Transaction not open! By default, journaled event batching requires an open database transaction.
+        MSG
+
+        connection.add_transaction_record(self)
+        @active = true
+        @joinable = true
+      end
+
+      def active?
+        @active
+      end
+
+      def joinable?
+        @joinable
+      end
+
+      def staged_events
+        @staged_events ||= []
+      end
+
+      # The following methods adhere to the API contract defined by:
+      # https://github.com/rails/rails/blob/v6.0.4.7/activerecord/lib/active_record/transactions.rb
+      #
+      # This allows our TransactionHandler to act as a "transaction record" and
+      # run callbacks before/after commit (or after rollback).
+      def before_committed!
+        Writer.enqueue!(*staged_events)
+        @joinable = false
```

Comment:
After we've flushed the events into the queue, we mark this transaction as non-joinable, but keep it "active" until the commit (or rollback) is fully completed.

This addresses an issue I found [(covered by this test case)](https://github.com/Betterment/journaled/pull/25/files#diff-d150802de35523c1e68d9701f8668a124dcfb9fc5c9e23c197442b7c449e653aR294) where if you try to journal an event in some other `before_commit` context, that thing might stage an event _after_ we've already flushed the staged events to the queue.

We definitely don't want to lose events, so instead the writer will see that the connection is no longer "joinable" and will directly enqueue those extra `before_commit` events.

Reply:
Could this be a common enough pattern to need a second batch? Is there any other way to ensure we come after any other journal enqueue?

Reply:
Yeah my first stab at solving this involved trying to get a journaled-managed transaction record in the "records" queue as early as possible, but I wasn't finding an obvious path forward there. I tried doing it as soon as the connection comes off the pool (i.e. gets checked out by the rails thread), but that's too early because there is no "joinable" transaction at that point.

So instead I tried walking up the list of transactions and (via some cracking open of internal ivars) inserting the txn record as early as I could in the earliest joinable transaction. It could work, but it felt a lot more fragile, and would break if you're intentionally using save-state-enabled nested transactions.

There's maybe a way to handle that case too, but it felt really fragile at that point -- it was way more complicated than just using the blessed API and letting Rails stick our "record" object on the queue.

Reply:
There's a case where I don't think this quite works. Because we add the journaled record to the transaction the first time we emit an event, if the first event emitted is within a before_commit hook, then it will get added to the records while the before_committed! methods are already being run for the transaction. Because the handler hasn't been processed yet, it's still "joinable" so the event is staged on it, but it's `before_committed!` method is never run, and the event is never sent.

I've been thinking about how to possibly fix this, and all I can think of is to stop using the transaction records completely and more directly monkey patching the Transaction or TransactionManager class to send them after the before hooks have ran.

Here is a spec I wrote up that tests this case and currently fails: [spec](https://github.com/smudge/journaled/compare/event-batching...medlefsen:journaled:me-event-batching#diff-d150802de35523c1e68d9701f8668a124dcfb9fc5c9e23c197442b7c449e653aR335-R347)